### PR TITLE
Fix broken behavior of IsVisibleInDynamoLibrary attribute with interfaces and Enums

### PR DESCRIPTION
**Pull request by aparajit-pratap**
### Purpose
https://jira.autodesk.com/browse/QNTM-4094

Bug description: When a class implements an interface method, although it is marked hidden, the interface method still appears as a node. 

This is a regression in 2.0 and can be traced back to language changes to fix polymorphism. In[ this change ](https://github.com/DynamoDS/Dynamo/pull/8626/commits/ade49d1461ef7f9d645419d692334b97f4297872) polymorphism support was added to interfaces where the interface method was allowed to accept a derived type and the node would at runtime resolve to the concrete implementation of the method matching the derived type.

This fix is to hide the interface methods altogether to address this issue as well as to remove any confusion of having extraneous interface nodes.

When attribute `IsVisibleInDynamoLibrary(false)` is defined on a class the following should be the expectation:
- [x] The current class is hidden from both library and search
- [x] Interface nodes should always be hidden (even if they are implemented by derived classes) as they do not have any implementation and their derived class implementations would be exposed as nodes if needed

The second issue is with the `IsVisibleInDynamoLibrary(false)` attribute not taking effect on an enum, which is a regression from 1.3.X. This can be traced back to the change made [here ](https://github.com/DynamoDS/Dynamo/pull/8311). This too has been fixed, which includes the change to not have to individually set the attributes on each enum field. The attributes, `IsVisibleInDynamoLibrary` and `Obsolete`  can now be applied globally on the enum, which now takes effect on all of the fields as well.

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@QilongTang

### CODE REVIEW

File: test/Engine/FFITarget/CodeCompletionClass.cs [hunk 1]
```diff
@@ -125,4 +126,16 @@ public class AnotherClassWithNameConflict
             public static string PropertyF { get; set; }
         }
     }
+
+    [IsVisibleInDynamoLibrary(false)]
+    [Serializable]
+    public class TraceableId : ISerializable
```
Comment (QilongTang) [inline test/Engine/FFITarget/CodeCompletionClass.cs]: @aparajit-pratap Can you explain how this test works, I can't see how adding a class with the tag verify stuff from my knowledge.

File: src/Engine/ProtoCore/FFI/CLRDLLModule.cs [hunk 2]
```diff
@@ -354,6 +354,10 @@ private ClassDeclNode ParseSystemType(Type type, string alias)
                     CLRModuleType.GetInstance(interf, Module, string.Empty);
                 }
             }
+            if(classnode.IsInterface)
```
Comment (QilongTang) [inline src/Engine/ProtoCore/FFI/CLRDLLModule.cs]: Are you able to track down why this got regressed? I do not see this check existed in previous version of code

File: test/Engine/FFITarget/CodeCompletionClass.cs (hunk 1 above)
Comment (aparajit-pratap) [reply test/Engine/FFITarget/CodeCompletionClass.cs]: @QilongTang I need to think how I can test for interfaces not appearing in search and library.

Comment (QilongTang) [reply test/Engine/FFITarget/CodeCompletionClass.cs]: @aparajit-pratap What do you think of either screen comparison or search results comparison?

**Comment by QilongTang** [comment]
LGTM

**Comment by QilongTang** [comment]
@aparajit-pratap Considering this bug should be there in Dynamo 2.0.2 as well. Do you mind cherry-picking to RC2.0.2 as well?

File: src/Engine/ProtoCore/FFI/CLRDLLModule.cs [hunk 3]
```diff
@@ -1268,24 +1272,45 @@ public FFIMethodAttributes(FieldInfo f)
...
+
             foreach (var attr in atts)
             {
                 //Set the obsolete message for enum fields.
-                if (attr is IsObsoleteAttribute)
+                if (attr is ObsoleteAttribute)
                 {
                     HiddenInLibrary = true;
-                    ObsoleteMessage = (attr as IsObsoleteAttribute).Message;
+                    ObsoleteMessage = (attr as ObsoleteAttribute).Message;
```
Comment (mjkkirschner) [inline src/Engine/ProtoCore/FFI/CLRDLLModule.cs]: @aparajit-pratap what is the difference between these attribute types?

**Pull request merged**